OpenRA.Mods.Cnc/Traits/Conditions/GrantConditionOnInfiltration.cs

```diff
@@ -0,0 +1,86 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Grants a condition when this building is infiltrated.")]
+	class GrantConditionOnInfiltrationInfo : ConditionalTraitInfo
+	{
+		public readonly BitSet<TargetableType> Types;
+
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		public readonly string Condition = null;
+
+		public readonly int Duration = 0;
+
+		public readonly bool ShowSelectionBar = true;
+		public readonly Color SelectionBarColor = Color.DarkBlue;
+
+		public override object Create(ActorInitializer init) { return new GrantConditionOnInfiltration(this); }
+	}
+
+	class GrantConditionOnInfiltration : ConditionalTrait<GrantConditionOnInfiltrationInfo>, INotifyInfiltrated, INotifyCreated, ITick, ISelectionBar
```

Comment:
Can we please make this trait interact with `TimedConditionBar` instead of duplicating the bar code here?

Reply:
That looks quite ugly, but I'll see what can I do.

Reply:
Should be straightforward - query and cache the traits implementing the notification interfaces, then ping them each tick.